### PR TITLE
feat: Add classifier explainability based on token importance

### PR DESCRIPTION
**Pull request by Pringled**
This PR adds explainability for classifiers by sorting the tokens of the input based on output layer logits for the predicted class. Couple of open questions/concerns:
- I feel like this adds even more functions (on top of the evaluate functions) to inference/model.py Is this the best place to add functions that are shared between inference/model.py and train/classifier.py, or should we have a utils.py or something similar? In the inference module makes sense since that's also installed when installing the train module, but perhaps a different file for functions that are shared between the two makes sense.
- Right now, if token_logits do not exist, they are computed when `get_most_important_tokens` is called. I think this is fine since it's really fast, but should we log this, or is it ok as is?
- typing for the model is set to Any since we get some nasty circular import stuff if we want to type it correctly, but maybe the lord of the types knows how to fix this 👀

### CODE REVIEW

**Review by stephantul — COMMENTED** [review]
I think the implementation contains a bug wrt the tokenization, you retokenize tokens. 

Here's what you could do: both the pipeline and the model already contain embeddings for all tokens separately, it's just the embedding matrix itself.

So getting token logits is in both cases just (paraphrased).

```python
model.model_head.predict_logits(model.model.embeddings)
```

No need to tokenize anything. So this is what I would start with, and then we'll see.

File: model2vec/inference/model.py [hunk 1]
```diff
@@ -305,3 +348,46 @@ def evaluate_single_or_multi_label(
...
+
+    # Determine the predicted label for the text
+    probs = model.predict_proba([text])[0]
+    label_idx = int(np.argmax(probs))
+
+    results = []
+    for token_id in unique_ids:
+        # Get the token string and logit
+        token_str = model.tokenizer.id_to_token(token_id)
+        token_logit = model.token_logits_cache.get(token_id)
```
Comment (stephantul) [inline model2vec/inference/model.py:384]: can this ever be `None`?

File: model2vec/inference/model.py [hunk 2]
```diff
@@ -305,3 +348,46 @@ def evaluate_single_or_multi_label(
...
+        if token_logit is None:
+            continue
+        # Get the logit for the predicted label
+        score = float(token_logit[label_idx])
```
Comment (stephantul) [inline model2vec/inference/model.py:388]: This looks kind of odd to me. Why would the logit of the predicted label, and not the collected logits of the token over all classes, determine the classification. For example, consider a situation in which a token gets high logits for 2 out of many classes. In that case ,it might not be important at all, right?

File: model2vec/inference/model.py [hunk 3]
```diff
@@ -305,3 +348,46 @@ def evaluate_single_or_multi_label(
...
+        results.append((token_str, score))
+
+    # Sort tokens by descending score
+    results.sort(key=lambda x: x[1], reverse=True)
```
Comment (stephantul) [inline model2vec/inference/model.py:392]: Could be nicer to not do an in-place sort.

File: model2vec/inference/model.py [hunk 4]
```diff
@@ -305,3 +348,46 @@ def evaluate_single_or_multi_label(
...
+    """
+    # Use the model's tokenizer to convert the text to token IDs
+    if isinstance(model, StaticModelPipeline):
+        input_ids = model.model.tokenize([text])
+    else:
+        input_ids = model.tokenize([text]).tolist()
+    unique_ids = set(input_ids[0])
+
+    # Identify tokens that are not yet cached and compute their logits
+    tokens_to_compute = [token_id for token_id in unique_ids if token_id not in model.token_logits_cache]
```
Comment (stephantul) [inline model2vec/inference/model.py:369]: maybe you can just do this in the loop below? For each token, you just get it from the cache, if you miss, you compute it? Just a small idea

File: model2vec/train/classifier.py [hunk 5]
```diff
@@ -246,6 +267,15 @@ def evaluate(
 
         return report
 
+    def get_most_important_tokens(self, text: str) -> list[tuple[str, float]]:
```
Comment (stephantul) [inline model2vec/train/classifier.py:270]: I think you can leave these functions out? I don't really see why they exist, except that they strongly couple the modules, where they could be decoupled before. i.e., if this function doesn't exist, there is no need to ever change this code if the explainability module changes, but now there suddenly is.

File: model2vec/inference/model.py [hunk 6]
```diff
@@ -192,6 +214,27 @@ def evaluate(
...
+        token_embeddings = self.embeddings[token_ids]
+        mlp = self.head[-1]
+        original_activation = mlp.out_activation_
+        # Set the activation function to identity to get the logits
+        mlp.out_activation_ = "identity"
+        logits = self.head.predict_proba(token_embeddings)
+        mlp.out_activation_ = original_activation
+        return logits
+
+    def get_most_important_tokens(self, text: str) -> list[tuple[str, float]]:
```
Comment (stephantul) [inline model2vec/inference/model.py:229]: idem as in the training code.

File: model2vec/inference/model.py [hunk 7]
```diff
@@ -192,6 +214,27 @@ def evaluate(
 
         return report
 
+    def predict_logits(self, token_ids: list[int]) -> np.ndarray:
+        """Predict the logits for the specified token IDs."""
+        # Extract embeddings for the specified token IDs.
+        token_embeddings = self.embeddings[token_ids]
+        mlp = self.head[-1]
```
Comment (stephantul) [inline model2vec/inference/model.py:221]: just a fair warning that this doesn't work if the head is ever not a pipeline with an mlp as final estimator. Which we do support, in principle.

**Review by stephantul — APPROVED** [review]
Just some nitpicks, I'm curious why you only check the logits for the predicted class, and not all classes. It seems to me that that could or should be part of what it means for a classifier to be explainable.
